tools/apidiff/cmd/changelog.go

```diff
@@ -0,0 +1,157 @@
+// Copyright 2018 Microsoft Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"errors"
+	"fmt"
+	"sort"
+	"strings"
+
+	"github.com/spf13/cobra"
+)
+
+var changelogCmd = &cobra.Command{
+	Use:   "changelog <package search dir> <base commit> <target commit>",
+	Short: "Generates a CHANGELOG report in markdown format for the packages under the specified directory.",
+	Long: `The changelog command generates a CHANGELOG for all of the packages under the directory specified in <package dir>.
+A table for added, removed, updated, and breaking changes will be created as required.`,
+	RunE: func(cmd *cobra.Command, args []string) error {
+		return theChangelogCmd(args)
+	},
+}
+
+func init() {
+	rootCmd.AddCommand(changelogCmd)
+}
+
+func theChangelogCmd(args []string) error {
+	// there should be exactly three args, a directory and two commit hashes
+	if len(args) < 3 {
+		return errors.New("not enough args were supplied")
+	} else if len(args) > 3 {
+		return errors.New("too many args were supplied")
+	} else if strings.Index(args[2], ",") > -1 {
+		return errors.New("sequence of target commits is not supported")
+	}
+
+	rpt, err := thePackagesCmd(args)
```

Comment:
I'm not crazy about this pattern, where commands end up calling each-other directly. It forces us to do argument checking in the body of this method, and it removes some strong typing that we could take advantage of in terms of what you need to execute.

Reply:
Agreed.  I was in a bit of a hurry and didn't want to refactor the `packages` command a whole bunch.  :/  I will look at cleaning this up in a separate PR.

Reply:
No worries, I totally get it. :)